jupyterhub/spawner.py

```diff
@@ -287,16 +306,77 @@ def _default_oauth_scopes(self):
         [Callable(), List()],
         help="""Allowed roles for oauth tokens.
 
+        Deprecated in 3.0: use oauth_client_allowed_scopes
+
         This sets the maximum and default roles
         assigned to oauth tokens issued by a single-user server's
         oauth client (i.e. tokens stored in browsers after authenticating with the server),
         defining what actions the server can take on behalf of logged-in users.
 
+        Default is an empty list, meaning minimal permissions to identify users,
+        no actions can be taken on their behalf.
+        """,
+    ).tag(config=True)
+
+    oauth_client_allowed_scopes = Union(
+        [Callable(), List()],
+        help="""Allowed scopes for oauth tokens issued by this server's oauth client.
+
+        This sets the maximum and default scopes
+        assigned to oauth tokens issued by a single-user server's
+        oauth client (i.e. tokens stored in browsers after authenticating with the server),
+        defining what actions the server can take on behalf of logged-in users.
+
         Default is an empty list, meaning minimal permissions to identify users,
         no actions can be taken on their behalf.
     """,
     ).tag(config=True)
 
+    def _get_oauth_client_allowed_scopes(self):
+        """Private method: get oauth allowed scopes
+
+        Handle:
+
+        - oauth_client_allowed_scopes
+        - callable config
+        - deprecated oauth_roles config
+        - access_scopes
+        """
+        # cases:
+        # 1. only scopes
+        # 2. only roles
+        # 3. both! (conflict, favor scopes)
+        scopes = []
+        if self.oauth_client_allowed_scopes:
+            allowed_scopes = self.oauth_client_allowed_scopes
+            if callable(allowed_scopes):
+                allowed_scopes = allowed_scopes(self)
```

Comment:
Is it obvious that the Callable in `oauth_client_allowed_scopes` must be a normal sync function?

I figure either declaring that assumption in the docstring or making us accept either could be relevant.

Reply:
Not obvious, so I made sure it now supports async (and stated as much in the help string)